common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java

```diff
@@ -0,0 +1,258 @@
+
+import static gov.cms.ab2d.common.util.Constants.AB2D_EPOCH;
+import static java.util.stream.Collectors.toList;
+
+@Slf4j
+@Service
+@Transactional
+public class CoverageServiceImpl implements CoverageService {
+
+    @Autowired
+    private CoverageRepository coverageRepo;
+
+    @Autowired
+    private CoveragePeriodRepository coveragePeriodRepo;
+
+    @Autowired
+    private CoverageSearchEventRepository coverageSearchEventRepo;
+
+    @Override
+    public CoveragePeriod getCoveragePeriod(long contractId, int month, int year) {
+        checkMonthAndYear(month, year);
+        return coveragePeriodRepo.getByContractIdAndMonthAndYear(contractId, month, year);
+    }
+
+    @Override
+    public boolean isCoveragePeriodInProgress(int searchId) {
+        CoveragePeriod period = coveragePeriodRepo.getOne(searchId);
+        JobStatus jobStatus = period.getStatus();
+        return jobStatus == JobStatus.IN_PROGRESS;
+    }
+
+    @Override
+    public boolean canEOBSearchBeStarted(int searchId) {
+        return !isCoveragePeriodInProgress(searchId);
+    }
+
+    @Override
+    public JobStatus getSearchStatus(int searchId) {
+
+        CoveragePeriod coverageSearch = coveragePeriodRepo.getOne(searchId);
+        return coverageSearch.getStatus();
+
+    }
+
+    @Override
+    public Optional<CoverageSearchEvent> findLastEvent(int searchId) {
+        CoveragePeriod period = coveragePeriodRepo.getOne(searchId);
+        return findLastEvent(period);
+    }
+
+    private Optional<CoverageSearchEvent> findLastEvent(CoveragePeriod period) {
+        return coverageSearchEventRepo.findFirstByCoveragePeriodOrderByCreatedDesc(period);
+    }
+
+    private CoverageSearchEvent getLastEvent(CoveragePeriod period) {
+        return findLastEvent(period).orElse(null);
+    }
+
+    @Transactional
+    @Override
+    public CoverageSearchEvent insertCoverage(int searchId, long searchEventId, Collection<String> beneficiaryIds) {
+
+        CoveragePeriod period = coveragePeriodRepo.getOne(searchId);
+        CoverageSearchEvent searchEvent = coverageSearchEventRepo.getOne(searchEventId);
+
+        Collection<Coverage> coverages = beneficiaryIds.stream()
+                .map(bene -> {
+                    Coverage coverage = new Coverage();
+                    coverage.setCoveragePeriod(period);
+                    coverage.setCoverageSearchEvent(searchEvent);
+                    coverage.setBeneficiaryId(bene);
+
+                    return coverage;
+                }).collect(toList());
+
+        JobStatus status = getSearchStatus(searchId);
+        if (status == JobStatus.IN_PROGRESS) {
+            coverageRepo.saveAll(coverages);
+        } else {
+            throw new InvalidJobAccessException("cannot update coverages if coverage search is not in progress (current status " + status + " )");
+        }
+
+        return searchEvent;
+    }
+
+    @Transactional
+    @Override
+    public void deletePreviousSearch(int searchId) {
```

Comment:
Where does this get called from?

Reply:
@adaykin this is actually up for debate. Basically we want all the previous search results in the database until the newest search has completed. The reason we want this is to diff our results to log any changes. Laila wants to log those changes so we can get a handle on when.

The thread actually doing the search could do these operations or deletePreviousSearch could be called in `CoverageServiceImpl.completeCoverageSearch()`

Reply:
Ok so it's planned to be added in the future? Just wanted to make sure you hadn't forgotten to call it somewhere.